crates/swc_ecma_minifier/src/pass/mangle_names/mod.rs

```diff
@@ -1,40 +1,291 @@
+            freq: &mut freq,
+            preserved,
+        });
+
+        freq
+    }
+
+    pub fn compile(self) -> Base54Chars {
+        static BASE54_DEFAULT_CHARS: &[u8; 64] =
+            b"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ$_0123456789";
+
+        let mut arr = BASE54_DEFAULT_CHARS
+            .iter()
+            .copied()
+            .enumerate()
+            .map(|(idx, c)| (self.0[idx], c))
+            .collect::<Vec<_>>();
+
+        arr.sort_by_key(|&(freq, _)| Reverse(freq));
+
+        let mut head = vec![];
+        let mut tail = vec![];
+
+        for (_, c) in arr {
+            if !(b'0'..=b'9').contains(&c) {
+                head.push(c);
+            }
+            tail.push(c);
+        }
+
+        Base54Chars {
+            head: head.try_into().unwrap(),
+            tail: tail.try_into().unwrap(),
+        }
+    }
+}
+
+struct CharFreqAnalyzer<'a> {
+    freq: &'a mut CharFreq,
+    preserved: &'a FxHashSet<Id>,
+}
+
+impl Visit for CharFreqAnalyzer<'_> {
+    noop_visit_type!();
+
+    visit_obj_and_computed!();
+
+    fn visit_ident(&mut self, i: &Ident) {
+        // It's not mangled
+        if self.preserved.contains(&i.to_id()) {
+            return;
+        }
+
+        self.freq.scan(&i.sym, -1);
+    }
+
+    fn visit_str(&mut self, s: &Str) {
+        self.freq.scan(&s.value, -1);
+    }
+
+    fn visit_prop_name(&mut self, n: &PropName) {
+        match n {
+            PropName::Ident(_) => {}
+            PropName::Str(_) => {}
+            PropName::Num(_) => {}
+            PropName::Computed(e) => e.visit_with(self),
+            PropName::BigInt(_) => {}
+        }
+    }
+
+    /// This is preserved anyway
+    fn visit_module_export_name(&mut self, _: &ModuleExportName) {}
+}
+
+impl AddAssign for CharFreq {
+    fn add_assign(&mut self, rhs: Self) {
+        for i in 0..64 {
+            self.0[i] += rhs.0[i];
+        }
+    }
+}
+
+impl Base54Chars {
+    /// givin a number, return a base54 encoded string
+    /// `usize -> [a-zA-Z$_][a-zA-Z$_0-9]*`
+    pub(crate) fn encode(&self, init: &mut usize, skip_reserved: bool) -> JsWord {
+        let mut n = *init;
+
+        *init += 1;
+
+        let mut base = 54;
+
+        while n >= base {
+            n -= base;
+            base <<= 6;
+        }
+
+        // Not sure if this is ideal, but it's safe
+        let mut ret: ArrayVec<_, 14> = ArrayVec::new();
+
+        base /= 54;
+        let mut c = self.head[n / base];
+        ret.push(c);
+
+        while base > 1 {
+            n %= base;
+            base >>= 6;
+            c = self.tail[n / base];
+
+            ret.push(c);
+        }
+
+        let s = unsafe {
+            // Safety: We are only using ascii characters
+            // Safety: The stack memory for ret is alive while creating JsWord
+            JsWord::from(std::str::from_utf8_unchecked(&ret))
+        };
+
+        if skip_reserved
+            && (s.is_reserved()
```

Comment:
We decide whether to skip the string only after it has been generated, which may not perform as well as before.

Reply:
Yeah it's one of the problem but seems like gzip size diff is quite big

Reply:
I agree with you.